### PR TITLE
Implement BIO.parMap2 ... parMap6 and parZip

### PR DESCRIPTION
**Pull request by Executioner1939**
#26 

@Avasil this one is ready for review :-).

### CODE REVIEW

**Comment by Executioner1939** [comment]
@Avasil Hi 😄 

Anything I can do still? Please let me know 😄 😄

File: core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    val ex = DummyException("dummy")
+    val f = BIO.parZip2(BIO(throw ex).delayExecution(1.second), BIO(2).delayExecution(2.seconds)).runToFuture
+
+    s.tick(1.second)
+    assertEquals(f.value, Some(Failure(ex)))
+  }
+
+  test("BIO.parZip2 should onError from the source before other and return the user error") { implicit s =>
+    val ex0 = DummyException("dummy")
+    val ex1 = DummyErrorResponse("dummy")
```
Comment (Avasil) [inline core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:88]: I feel like `DummyErrorResponse` class is unnecessary, we can use `DummyException` with different message instead

File: core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    val f = BIO
+      .parZip2(BIO(throw ex0).mapError(_ => ex1).delayExecution(2.second), BIO(2).delayExecution(1.seconds))
```
Comment (Avasil) [inline core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:91]: IMO `mapError`s in these tests are not relevant to `parZip` and could be removed

Comment (Executioner1939) [reply core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:91]: Hi Avasil :-). I see I completely misunderstood, I thought thow would result in a terminal error. This test is redundant then.

I have gone through the runloop and seen that the Eval and Suspend Nodes catch the Error and raise it to the non-terminal error channel.

File: core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala (hunk 1 above)
Comment (Executioner1939) [reply core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala:88]: Agreed, I will remove this :-).

**Comment by Executioner1939** [comment]
@Avasil I updated the PR, removed the `DummyErrorResponse`, removed the tests using `mapError` and added two using `BIO.terminate`, one for each "side" of `parZip2`.

File: core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala (hunk 2 above)
Comment (Avasil) [reply core/shared/src/test/scala/monix/bio/BIOParZipSuite.scala]: `BIO.apply` which is alias to `eval` catches any errors and that's why it returns `BIO[Throwable, A]`

It doesn't happen for other methods, e.g. `flatMap(_ => throw ex)` won't be caught in typed error channel.

There is also `evalTotal` (and UIO.apply) which returns `UIO` so any error thrown there would be terminal

**Pull request merged**